Broad api calls

xtrojak opened the pull request:
Improvement of API calls efficiency - where possible, broader calls were created (instead of obtaining single result, e.g. InChIKey, we obtain multiple attributes at once). Results are stored in cache for the single spectra data, as the obtained result can be reused for another job. This was implemented in `Annotator.execute_job_with_cache`.

Another `cache` for async requests was implemented using asynchronous version of [lru_cache](https://asyncstdlib.readthedocs.io/en/latest/source/api/functools.html#asyncstdlib.functools.lru_cache). Close #8.

Additionally, implementation of individual services was simplified where possible (mostly by separating calls and parsing), but this caused issue #19.

hechth commented on libs/utils/HashableDict.py:
```diff
@@ -0,0 +1,3 @@
+class HashableDict(dict):
+    def __hash__(self):
+        return hash(tuple(sorted(self.items())))
```
Isn't that already a class in some base library? Or what is special about it?

xtrojak replied:
Yes, good points, replaced with `frozendict` in e046b2a (the lib is on conda).

hechth reviewed (commented):
Code looks good, only problem is that asyncstdlib is not on conda!

martenson reviewed (approved):
Seems fine. I've added simple test data input and output files. 

The efficiency of the double caching with both `execute_job_with_cache` and the `@lru_cache` method decorator is unclear to me. Testing with a significant dataset could clarify this, but is not required at the moment in my opinion.

Beware that the default lru_cache maxsize is unlimited.

martenson commented:
Oh, and we have a 500MB .msp file in the UMSA library for some serious testing: https://umsa.cerit-sc.cz/library/list#folders/F1c84aa7fc4490e6d/datasets/9a34fd777b6c8572

martenson commented on libs/services/Converter.py:
```diff
@@ -6,6 +9,7 @@ class Converter:
     def __init__(self, session):
         self.session = session
 
+    @lru_cache
```
Are you sure `self` is hashable here? It contains the session.

xtrojak replied:
You are right, we have to make sure default hash based on `id` is sufficient here. We will have a look into that in #1 (cache testing).

xtrojak commented:
> Seems fine. I've added simple test data input and output files.
> 
> The efficiency of the double caching with both `execute_job_with_cache` and the `@lru_cache` method decorator is unclear to me. Testing with a significant dataset could clarify this, but is not required at the moment in my opinion.

The current design requires specification of type `(source, target, service)`, which defines that user wants to obtain `target` attribute based on `source` attribute using particular `service`. In this PR, I have changed queries to obtain as much data as possible from single API call. In `execute_job_with_cache` we cache these data for single spectra. Then for another triple requesting an already cached attribute we can just return it. Note that the request itself would be actually different from the previous one, therefore `lru_cache` would not help.

`lru_cache`, on the other hand, is used to avoid executing the exactly same requests multiple times. Mostly applied when spectra in a single `.msp` file share an attribute (which seems to be the case).


> 
> Beware that the default lru_cache maxsize is unlimited.

It seems like default is 128 ([default arg value](https://asyncstdlib.readthedocs.io/en/latest/source/api/functools.html#asyncstdlib.functools.lru_cache))
